enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,229 @@
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will include the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There will be no explicit dependency on the `greenboot` RPM.
+The script will be installed into the `/etc/greenboot/check/required.d` directory
+and it will not be executed during the system boot in case the `greenboot` package
+is not present.
+
+> If other scripts are present in the `/etc/greenboot/check/required.d` directory,
+> they will be executed by `greenboot` in the alphanumeric order.
+
+The script performs the following sanity checks:
+* Exit if the current user is not `root`
+* Exit if MicroShift service is not enabled
+* Wait for the MicroShift service to become active
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+The health check script will implement validations for the core MicroShift state
+and its services running as pods in the following namespaces:
+- openshift-ovn-kubernetes
+- openshift-service-ca
+- openshift-ingress
+- openshift-dns
+- openshift-storage
+
+> When pluggable CNI and CSI functionality is implemented, the `openshift-ovn-kubernetes`
+> and `openshift-storage` namespace name selection need to be conditional on
+> the loaded CNI and CSI plugins. The plugin developers may choose to add plugin-
+> specific health check scripts.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+Greenboot redirects all the script output to the system log, accessible via the
+`journalctl -u greenboot-healthcheck.service` command.
+
+|Validation                                        |Pass  |Fail  |
+|--------------------------------------------------|------|------|
+|Check microshift.service is enabled               |Next  |exit 0|
+|Wait for microshift.service to be active (!failed)|Next  |exit 1|
+|Wait for any pods to start                        |Next  |exit 1|
+|For each core namespace, wait for pods to be ready|Next  |exit 1|
+|For each core namespace, check pods not restarting|exit 0|exit 1|
+
+> If the system is not booted using the `ostree` file system, the health check
+> procedures still run, but no rollback would be possible in case of an upgrade
+> failure.
+
+The wait timeout of each validation starts from the base of 5 minutes and it is
```

Comment:
How was the default timeout chosen?

Reply:
Empirically. If it's too low for certain cases, it is increased with every unsuccessful boot.